[release/3.1] Add documentation for packaging source-build (#1653)
dotnet/source-build: Documentation/packaging-installation.md

# .NET Core Packaging and Installation

This document helps you install or package an SDK built using source-build.

The SDK built by source-build is generally not portable. That means it will work on the same operating system where it was built. It will not work on older operating systems. It may work on newer operating systems.

## Using the SDK directly

The built SK is generally located at `artifacts/${ARCHITECTURE}/Release/dotnet-sdk-${SDK_VERSION}-${RUNTIME_ID}.tar.gz`. To install and use it, unpack it to any directory and then use the `dotnet` executable.

For example, you can extract the tarball to `/opt/dotnet` and use it from there:

```
$ sudo mkdir /opt/dotnet/
$ sudo tar xf artifacts/x64/Release/dotnet-sdk-3.1.105-fedora.32.tar.gz -C /opt/dotnet/
$ /opt/dotnet/dotnet --info

> ```

Author: I have split the document into 3 sections: 

- installing per-user
- installing system-wide
- making a distro package

1. Extract the tarball to the distribution appropriate location such as `/usr/lib64/dotnet/`, `/usr/lib/dotnet/`, or `/usr/lib/x86_64-linux-gnu/dotnet/`

2. Create a symlink from `/usr/bin/dotnet` (or equivalent) to where you installed the SDK in the previous step. This lets users simply run `dotnet` and have it work.

3. Create an `/etc/dotnet/install_location` and add the location where you installed the SDK in there. This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).

4. Define the environment variable `DOTNET_ROOT` and make it point to where you installed the SDK. This environment variable is used instead of `/etc/dotnet/install_location` by older .NET Core versions.

5. Add `~/.dotnet/tools` to the `$PATH`. Otherwise you wont be able to run dotnet tools that you install directly.

There are several other optional things you can do:

- .NET Core source repositories include man pages. You can search for them and package them up: `find -iname '*.1' -exec cp {} /usr/share/man/man1/ \;`

> - .NET Core includes bash-completion and zsh-completion scripts. You can install them.

Author: I guess it speaks to your suggestion on linking to a better resource, because that page should be replaced by https://github.com/dotnet/cli/pull/13384

> - .NET Core includes bash-completion and zsh-completion scripts. You can install them.

Author: Done! Thanks for the idea.

# .NET Core Packaging and Installation

This document helps you install or package a .NET Core SDK built using source-build.

The SDK built by source-build is generally not portable. That means it will work on the same operating system where it was built. It will not work on older operating systems. It may work on newer operating systems.

The built SDK is generally located at `artifacts/${ARCHITECTURE}/Release/dotnet-sdk-${SDK_VERSION}-${RUNTIME_ID}.tar.gz`.

## Using the SDK directly (install per-user)

To use the SDK directly, unpack the SDK to any directory and then use the `dotnet` executable from it.

You can find more details on how to manually install an SDK from a tarball at [manually installing an SDK](https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#manual-install). The same steps listed should work for all Linux distributions where `bash` is the default shell.

## Installing the SDK globally

If you want to install the SDK globally (and not per user), here are some suggestions.

1. Extract the tarball to a distribution-appropriate location such as `/usr/local/lib64/dotnet/`, `/usr/local/lib/dotnet/`, or `/usr/local/lib/x86_64-linux-gnu/dotnet/`.

> 2. Create a symlink from `/usr/local/bin/dotnet` (or equivalent) to where you installed the SDK in the previous step. This lets users simply run `dotnet` and have it work.

Reviewer: Should this mention "to the `dotnet` executable in the location you installed the SDK"? I think I've seen some people tripped up by the path ending up with `.../dotnet/dotnet`. Maybe just give `/usr/local/lib64/dotnet/dotnet` as an example?

Author: Good idea! Thanks.

> 3. Create an `/etc/dotnet/install_location` and add the location where you installed the SDK in there. This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).

Reviewer: Seems worth mentioning that this needs to be the dir (contrasting vs. what you might expect from prev step).

> 3. Create an `/etc/dotnet/install_location` file and add the path of the SDK directory in there (such as `/usr/local/lib64/dotnet`). This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).

Reviewer: A bit more specific what should be in the file? (And "add" is a little confusing IMO since it's a new file, I usually look for an example/pattern to follow when I see "add".)

```suggestion
3. Create an `/etc/dotnet/install_location` file with a single line of content, the path of the SDK directory (such as `/usr/local/lib64/dotnet`). This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).
```

4. Define `DOTNET_ROOT` and update `PATH` by saving the following as `/etc/profile.d/dotnet-local.sh` (or equivalent)

   ```bash
   # Set location for AppHost lookup
   [ -z "$DOTNET_ROOT" ] && export DOTNET_ROOT=/usr/local/lib64/dotnet

   # Add dotnet tools directory to PATH
   DOTNET_TOOLS_PATH="$HOME/.dotnet/tools"
   case "$PATH" in
       *"$DOTNET_TOOLS_PATH"* ) true ;;
       * ) PATH="$PATH:$DOTNET_TOOLS_PATH" ;;
   esac
   ```

>    This allows apphost-lookup to work via `DOTNET_ROOT` and allows users to easily use dotnet tools directly after a `dotnet tool install`. This snippet only works on `bash`. You will need to adapt it for other shells.

Reviewer: The snippet looks posixy to me and seems to work in an Ubuntu container with `dash`. On the other hand, `zsh` doesn't load `profile.d` and has some other way of doing things entirely. `will need to adapt it` => `may need to adapt it`, or am I missing something?

(I might not have tested `dash` properly, not sure if it might have some compat modes on.)

Author: I think you are right. We wrote this snippet in a way to be explicitly compatible with `sh` (unless we introduced some bugs).

## Creating a Linux distribution package

If you want to create a Linux distribution package (`rpm`, `deb`) out of the source-build SDK here are some suggestions.

See [.NET Core distribution packaging](https://docs.microsoft.com/en-us/dotnet/core/distribution-packaging) for information on suggested packages, subpackages, name and contents.

This is the minimal amount of content you need to package up:

1. Extract the tarball to the distribution appropriate location such as `/usr/lib64/dotnet/`, `/usr/lib/dotnet/`, or `/usr/lib/x86_64-linux-gnu/dotnet/`.

> 2. Create a symlink from `/usr/bin/dotnet` (or equivalent) to the `dotnet` binary in SDK that you installed in the previous step. Foe example, `/usr/bin/dotnet` -> `/usr/lib64/dotnet/dotnet`. This lets users simply run `dotnet` and have it work.

Reviewer: ```suggestion
2. Create a symlink from `/usr/bin/dotnet` (or equivalent) to the `dotnet` binary in the SDK that you installed in the previous step. For example, `/usr/bin/dotnet` -> `/usr/lib64/dotnet/dotnet`. This lets users simply run `dotnet` and have it work.
```

> 3. Create an `/etc/dotnet/install_location` file and add the path of the SDK directory in there (such as `/usr/lib64/dotnet`). This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).

Reviewer: ```suggestion
3. Create an `/etc/dotnet/install_location` file with a single line of content, the path of the SDK directory (such as `/usr/lib64/dotnet`). This file is used by [.NET Core to find the SDK/Runtime location](https://github.com/dotnet/designs/blob/master/accepted/2020/install-locations.md).
```

Author: Thanks for raising this. I have added an "example" file to hopefully make it clearer to the reader.

Reviewer: I like it, thanks. 😄
